File: Java/06. Software Design Principles SOLID & More/Practice17.java
/**
 * Practice 17: Applying SRP (Single Responsibility Principle)
 * Task: Refactor OnDemandAgentService so each class has a single responsibility.
 *
 * How to compile and run:
 *   javac Practice17.java
 *   java Practice17
 */

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

public class Practice17 {

    static class Customer {
        String name;

        Customer(String name) {
            this.name = name;
        }
    }

    static class Agent {
        String name;
        boolean isAvailable;

        Agent(String name, boolean isAvailable) {
            this.name = name;
            this.isAvailable = isAvailable;
        }
    }

    static class AssignmentRepository {
        void save(String customerName, String agentName) {
            System.out.println("[DB] Assignment saved: customer=" + customerName + ", agent=" + agentName);
        }
    }

    static class NotificationService {
        void notifyCustomer(String customerName, String agentName) {
            System.out.println("[Notify] Dear " + customerName + ", " + agentName + " is on the way.");
        }

        void notifyAgent(String agentName, String customerName) {
            System.out.println("[Notify] Hi " + agentName + ", new trip assigned for " + customerName + ".");
        }
    }

    static class AuditLogger {
        void logAssignment(String customerName, String agentName) {
            System.out.println("[Audit] " + LocalDateTime.now() + " Assigned " + agentName + " to " + customerName);
        }
    }

    static class AgentMatcher {
        Agent findAvailableAgent(List<Agent> agents) {
            for (Agent agent : agents) {
                if (agent.isAvailable) {
                    return agent;
                }
            }
            throw new IllegalStateException("No available agent found.");
        }
    }

    static class OnDemandAgentService {
        private final AgentMatcher matcher;
        private final AssignmentRepository repository;
        private final NotificationService notifier;
        private final AuditLogger logger;

        OnDemandAgentService(
            AgentMatcher matcher,
            AssignmentRepository repository,
            NotificationService notifier,
            AuditLogger logger
        ) {
            this.matcher = matcher;
            this.repository = repository;
            this.notifier = notifier;
            this.logger = logger;
        }

        Agent assignAgent(Customer customer, List<Agent> agents) {
            Agent agent = matcher.findAvailableAgent(agents);
            agent.isAvailable = false;

            repository.save(customer.name, agent.name);
            notifier.notifyCustomer(customer.name, agent.name);
            notifier.notifyAgent(agent.name, customer.name);
            logger.logAssignment(customer.name, agent.name);

            return agent;
        }
    }

    public static void main(String[] args) {
        Customer customer = new Customer("Afsana");
        List<Agent> agents = new ArrayList<>();
        agents.add(new Agent("Robin", false));
        agents.add(new Agent("Pulok", true));

        OnDemandAgentService service = new OnDemandAgentService(
            new AgentMatcher(),
            new AssignmentRepository(),
            new NotificationService(),
            new AuditLogger()
        );

        Agent assigned = service.assignAgent(customer, agents);
        System.out.println("Result: " + assigned.name + " assigned successfully.");
    }
}

File: C#/06. Software Design Principles SOLID & More/Practice17.cs
/**
 * Practice 17: Applying SRP (Single Responsibility Principle)
 * Task: Refactor OnDemandAgentService so each class has a single responsibility.
 *
 * How to run (using dotnet-script):
 *   dotnet script Practice17.cs
 */

using System;
using System.Collections.Generic;

class Practice17
{
    class Customer
    {
        public string Name;

        public Customer(string name)
        {
            Name = name;
        }
    }

    class Agent
    {
        public string Name;
        public bool IsAvailable;

        public Agent(string name, bool isAvailable)
        {
            Name = name;
            IsAvailable = isAvailable;
        }
    }

    class AssignmentRepository
    {
        public void Save(string customerName, string agentName)
        {
            Console.WriteLine($"[DB] Assignment saved: customer={customerName}, agent={agentName}");
        }
    }

    class NotificationService
    {
        public void NotifyCustomer(string customerName, string agentName)
        {
            Console.WriteLine($"[Notify] Dear {customerName}, {agentName} is on the way.");
        }

        public void NotifyAgent(string agentName, string customerName)
        {
            Console.WriteLine($"[Notify] Hi {agentName}, new trip assigned for {customerName}.");
        }
    }

    class AuditLogger
    {
        public void LogAssignment(string customerName, string agentName)
        {
            Console.WriteLine($"[Audit] {DateTime.Now:O} Assigned {agentName} to {customerName}");
        }
    }

    class AgentMatcher
    {
        public Agent FindAvailableAgent(List<Agent> agents)
        {
            foreach (var agent in agents)
            {
                if (agent.IsAvailable)
                    return agent;
            }
            throw new InvalidOperationException("No available agent found.");
        }
    }

    class OnDemandAgentService
    {
        private readonly AgentMatcher matcher;
        private readonly AssignmentRepository repository;
        private readonly NotificationService notifier;
        private readonly AuditLogger logger;

        public OnDemandAgentService(
            AgentMatcher matcher,
            AssignmentRepository repository,
            NotificationService notifier,
            AuditLogger logger)
        {
            this.matcher = matcher;
            this.repository = repository;
            this.notifier = notifier;
            this.logger = logger;
        }

        public Agent AssignAgent(Customer customer, List<Agent> agents)
        {
            var agent = matcher.FindAvailableAgent(agents);
            agent.IsAvailable = false;

            repository.Save(customer.Name, agent.Name);
            notifier.NotifyCustomer(customer.Name, agent.Name);
            notifier.NotifyAgent(agent.Name, customer.Name);
            logger.LogAssignment(customer.Name, agent.Name);

            return agent;
        }
    }

    static void Main(string[] args)
    {
        var customer = new Customer("Afsana");
        var agents = new List<Agent>
        {
            new Agent("Robin", false),
            new Agent("Pulok", true)
        };

        var service = new OnDemandAgentService(
            new AgentMatcher(),
            new AssignmentRepository(),
            new NotificationService(),
            new AuditLogger());

        var assigned = service.AssignAgent(customer, agents);
        Console.WriteLine($"Result: {assigned.Name} assigned successfully.");
    }
}
